state/validatorsInfoMap.go

```diff
@@ -0,0 +1,224 @@
+
+	for shardID, valInShard := range input {
+		for _, val := range valInShard {
+			ret.valInfoMap[shardID] = append(ret.valInfoMap[shardID], val)
+		}
+	}
+
+	return ret
+}
+
+// GetAllValidatorsInfo returns a []ValidatorInfoHandler copy with validators from all shards.
+func (vi *shardValidatorsInfoMap) GetAllValidatorsInfo() []ValidatorInfoHandler {
+	ret := make([]ValidatorInfoHandler, 0)
+
+	vi.mutex.RLock()
+	defer vi.mutex.RUnlock()
+
+	for _, validatorsInShard := range vi.valInfoMap {
+		validatorsCopy := make([]ValidatorInfoHandler, len(validatorsInShard))
+		copy(validatorsCopy, validatorsInShard)
+		ret = append(ret, validatorsCopy...)
+	}
+
+	return ret
+}
+
+// GetShardValidatorsInfoMap returns a <shard, ValidatorInfoHandler> map copy of internally stored data
+func (vi *shardValidatorsInfoMap) GetShardValidatorsInfoMap() map[uint32][]ValidatorInfoHandler {
+	ret := make(map[uint32][]ValidatorInfoHandler, len(vi.valInfoMap))
+
+	vi.mutex.RLock()
+	defer vi.mutex.RUnlock()
+
+	for shardID, validatorsInShard := range vi.valInfoMap {
+		validatorsCopy := make([]ValidatorInfoHandler, len(validatorsInShard))
+		copy(validatorsCopy, validatorsInShard)
+		ret[shardID] = validatorsCopy
+	}
+
+	return ret
+}
+
+// Add adds a ValidatorInfoHandler in its corresponding shardID
+func (vi *shardValidatorsInfoMap) Add(validator ValidatorInfoHandler) error {
+	if check.IfNil(validator) {
+		return ErrNilValidatorInfo
+	}
+
+	shardID := validator.GetShardId()
+	vi.mutex.Lock()
+	vi.valInfoMap[shardID] = append(vi.valInfoMap[shardID], validator)
+	vi.mutex.Unlock()
+
+	return nil
+}
+
+// GetValidator returns a ValidatorInfoHandler copy with the provided blsKey,
+// if it is present in the map, otherwise returns nil
+func (vi *shardValidatorsInfoMap) GetValidator(blsKey []byte) ValidatorInfoHandler {
+	for _, validator := range vi.GetAllValidatorsInfo() {
+		if bytes.Equal(validator.GetPublicKey(), blsKey) {
+			return validator
+		}
+	}
+
+	return nil
+}
+
+// Replace will replace an existing ValidatorInfoHandler with a new one. The old and new validator
+// shall be in the same shard. If the old validator is not found in the map, an error is returned
+func (vi *shardValidatorsInfoMap) Replace(old ValidatorInfoHandler, new ValidatorInfoHandler) error {
+	if check.IfNil(old) {
+		return fmt.Errorf("%w for old validator in shardValidatorsInfoMap.Replace", ErrNilValidatorInfo)
+	}
+	if check.IfNil(new) {
+		return fmt.Errorf("%w for new validator in shardValidatorsInfoMap.Replace", ErrNilValidatorInfo)
+	}
+	if old.GetShardId() != new.GetShardId() {
+		return fmt.Errorf("%w when trying to replace %s from shard %v with %s from shard %v",
+			ErrValidatorsDifferentShards,
+			hex.EncodeToString(old.GetPublicKey()),
+			old.GetShardId(),
+			hex.EncodeToString(new.GetPublicKey()),
+			new.GetShardId(),
+		)
+	}
+
+	shardID := old.GetShardId()
+
+	vi.mutex.Lock()
+	defer vi.mutex.Unlock()
+
+	for idx, validator := range vi.valInfoMap[shardID] {
```

Comment:
slight optimization:
```GO
validatorSlice := vi.valInfoMap[shardID]
for idx, validator := range validatorSlice {
....
```

Reply:
Actually, it seems like this is slightly less efficient. I've run a benchmark on a map with 1,000,000 entries which
- `Iterate1` iterates using `for range vi.valInfoMap[shardID]`
- `Iterate2` iterates using:
``` go
 vSlice := vi.valInfoMap[shardID]
	for range vSlice {
}
```

Results:
```
BenchmarkShardValidatorsInfoMap_IterateMap1/iterate_1
BenchmarkShardValidatorsInfoMap_IterateMap1/iterate_1-8         	1000000000	         0.0002455 ns/op
BenchmarkShardValidatorsInfoMap_IterateMap1/iterate_2
BenchmarkShardValidatorsInfoMap_IterateMap1/iterate_2-8         	1000000000	         0.0002704 ns/op
```

Reply:
ok